File: apps/api/app/translation/languages.py
"""Language metadata for the 13 supported languages (PRD FR-12 / task
Section 2). A single source of truth for native/English names, script
direction, and the code list used to restrict language detection and
validate `language` fields elsewhere in the app.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class LanguageInfo:
    code: str
    native_name: str
    english_name: str
    direction: str  # "ltr" | "rtl"
    script: str
    supported: bool = True


LANGUAGES: dict[str, LanguageInfo] = {
    "en": LanguageInfo("en", "English", "English", "ltr", "Latin"),
    "hi": LanguageInfo("hi", "हिन्दी", "Hindi", "ltr", "Devanagari"),
    "mr": LanguageInfo("mr", "मराठी", "Marathi", "ltr", "Devanagari"),
    "bn": LanguageInfo("bn", "বাংলা", "Bengali", "ltr", "Bengali"),
    "ta": LanguageInfo("ta", "தமிழ்", "Tamil", "ltr", "Tamil"),
    "te": LanguageInfo("te", "తెలుగు", "Telugu", "ltr", "Telugu"),
    "gu": LanguageInfo("gu", "ગુજરાતી", "Gujarati", "ltr", "Gujarati"),
    "kn": LanguageInfo("kn", "ಕನ್ನಡ", "Kannada", "ltr", "Kannada"),
    "ml": LanguageInfo("ml", "മലയാളം", "Malayalam", "ltr", "Malayalam"),
    "pa": LanguageInfo("pa", "ਪੰਜਾਬੀ", "Punjabi", "ltr", "Gurmukhi"),
    "or": LanguageInfo("or", "ଓଡ଼ିଆ", "Odia", "ltr", "Odia"),
    "as": LanguageInfo("as", "অসমীয়া", "Assamese", "ltr", "Bengali"),
    "ur": LanguageInfo("ur", "اردو", "Urdu", "rtl", "Arabic"),
}

SUPPORTED_LANGUAGE_CODES: tuple[str, ...] = tuple(LANGUAGES.keys())

# Canonical pivot language: everything is translated to/from English before
# touching the existing RAG pipeline (task Section 4 - "do NOT create
# separate knowledge bases per language").
DEFAULT_LANGUAGE = "en"


def is_supported(code: str | None) -> bool:
    return bool(code) and code in LANGUAGES


def get_language(code: str) -> LanguageInfo:
    return LANGUAGES[code]

File: apps/web/src/api/languages.ts
// Mirrors apps/api/app/translation/languages.py - single source of truth
// for the 13 supported languages on the frontend side. Keep in sync if
// the backend list changes.

export type LanguageCode =
  | 'en'
  | 'hi'
  | 'mr'
  | 'bn'
  | 'ta'
  | 'te'
  | 'gu'
  | 'kn'
  | 'ml'
  | 'pa'
  | 'or'
  | 'as'
  | 'ur'

export interface LanguageInfo {
  code: LanguageCode
  nativeName: string
  englishName: string
  direction: 'ltr' | 'rtl'
}

export const LANGUAGES: LanguageInfo[] = [
  { code: 'en', nativeName: 'English', englishName: 'English', direction: 'ltr' },
  { code: 'hi', nativeName: 'हिन्दी', englishName: 'Hindi', direction: 'ltr' },
  { code: 'mr', nativeName: 'मराठी', englishName: 'Marathi', direction: 'ltr' },
  { code: 'bn', nativeName: 'বাংলা', englishName: 'Bengali', direction: 'ltr' },
  { code: 'ta', nativeName: 'தமிழ்', englishName: 'Tamil', direction: 'ltr' },
  { code: 'te', nativeName: 'తెలుగు', englishName: 'Telugu', direction: 'ltr' },
  { code: 'gu', nativeName: 'ગુજરાતી', englishName: 'Gujarati', direction: 'ltr' },
  { code: 'kn', nativeName: 'ಕನ್ನಡ', englishName: 'Kannada', direction: 'ltr' },
  { code: 'ml', nativeName: 'മലയാളം', englishName: 'Malayalam', direction: 'ltr' },
  { code: 'pa', nativeName: 'ਪੰਜਾਬੀ', englishName: 'Punjabi', direction: 'ltr' },
  { code: 'or', nativeName: 'ଓଡ଼ିଆ', englishName: 'Odia', direction: 'ltr' },
  { code: 'as', nativeName: 'অসমীয়া', englishName: 'Assamese', direction: 'ltr' },
  { code: 'ur', nativeName: 'اردو', englishName: 'Urdu', direction: 'rtl' },
]

const BY_CODE = new Map(LANGUAGES.map((l) => [l.code, l]))

export function getLanguage(code: LanguageCode): LanguageInfo {
  return BY_CODE.get(code) ?? LANGUAGES[0]
}

export function isRtl(code: LanguageCode): boolean {
  return getLanguage(code).direction === 'rtl'
}

export function isSupportedLanguage(code: string | null | undefined): code is LanguageCode {
  return !!code && BY_CODE.has(code as LanguageCode)
}
